lib/spack/spack/cmd/config.py

```diff
@@ -135,11 +164,150 @@ def config_list(args):
+            config_dict = data
+        # update all sections from config dict
+        # We have to iterate on keys to keep overrides from the file
+        for section in config_dict.keys():
+            if section in spack.config.section_schemas.keys():
+                # Special handling for compiler scope difference
+                # Has to be handled after we choose a section
+                if scope is None:
+                    if section == 'compilers':
+                        scope = spack.config.default_modify_scope()
+                    else:
+                        scope = spack.config.default_modify_scope(False)
+
+                value = config_dict[section]
+                existing = spack.config.get(section, scope=scope)
+                new = spack.config.merge_yaml(existing, value)
+
+                if re.match(r'env.*', scope or ''):
+                    e = ev.get_env(args, 'config add')
+                    e.set_config(section, new)
+                else:
+                    spack.config.set(section, new, scope=scope)
+
+    if args.value:
+        components = spack.config.process_config_path(args.value)
+
+        has_existing_value = True
+        path = ''
+        override = False
+        for idx, name in enumerate(components[:-1]):
+            # First handle double colons in constructing path
+            colon = '::' if override else ':' if path else ''
+            path += colon + name
+            if getattr(name, 'override', False):
+                override = True
+            else:
+                override = False
+
+            # Test whether there is an existing value at this level
+            existing = spack.config.get(path, scope=scope)
+
+            if existing is None:
+                has_existing_value = False
+                # We've nested further than existing config, so we need the
+                # type information for validation to know how to handle bare
+                # values appended to lists.
+                existing = spack.config.type_of(path)
+
+                # construct value from this point down
+                value = syaml.load_config(components[-1])
+                for component in reversed(components[idx + 1:-1]):
+                    value = {component: value}
+                break
+
+        if has_existing_value:
+            path, _, value = args.value.rpartition(':')
+            value = syaml.load_config(value)
+            existing = spack.config.get(path, scope=scope)
+
+        # append values to lists appropriately
+        if isinstance(existing, list) and not isinstance(value, list):
+            value = [value]
+
+        # merge value into existing
+        new = spack.config.merge_yaml(existing, value)
+
+        if re.match(r'env.*', scope or ''):
+            e = ev.get_env(args, 'config add')
+            e.set_config(path, new)
+        else:
+            spack.config.set(path, new, scope=scope)
```

Comment:
same question here as above -- also this logic is duplicated, which makes it hard to maintain.  At the very least this should be a function in the command.